pkg/appgateserver/cmd/cmd.go

```diff
@@ -0,0 +1,186 @@
+1. Each request contains the '?senderAddress=[address]' query parameter
+2. The key associated with the '--signing-key' flag belongs to the address
+   provided in the request, otherwise the ring signature will not be valid.`,
+		Args: cobra.NoArgs,
+		RunE: runAppGateServer,
+	}
+
+	cmd.Flags().StringVar(&flagSigningKey, "signing-key", "", "The name of the key that will be used to sign relays")
+	cmd.Flags().StringVar(&flagListeningEndpoint, "listening-endpoint", "http://localhost:42069", "The host and port that the appgate server will listen on")
+	cmd.Flags().StringVar(&flagCometWebsocketUrl, "comet-websocket-url", "ws://localhost:36657/websocket", "The URL of the comet websocket endpoint to communicate with the pocket blockchain")
+	cmd.Flags().BoolVar(&flagSelfSigning, "self-signing", false, "Whether the server should sign all incoming requests with its own ring (for applications)")
+
+	cmd.Flags().String(flags.FlagKeyringBackend, "", "Select keyring's backend (os|file|kwallet|pass|test)")
+	cmd.Flags().String(flags.FlagNode, "tcp://localhost:36657", "The URL of the comet tcp endpoint to communicate with the pocket blockchain")
+
+	return cmd
+}
+
+func runAppGateServer(cmd *cobra.Command, _ []string) error {
+	// Create a context that is canceled when the command is interrupted
+	ctx, cancelCtx := context.WithCancel(cmd.Context())
+	defer cancelCtx()
+
+	// Retrieve the client context for the chain interactions.
+	clientCtx := cosmosclient.GetClientContextFromCmd(cmd)
+
+	// Parse the listening endpoint.
+	listeningUrl, err := url.Parse(flagListeningEndpoint)
+	if err != nil {
+		return fmt.Errorf("failed to parse listening endpoint: %w", err)
+	}
+
+	log.Printf("INFO: Creating block client, using comet websocket URL: %s...", flagCometWebsocketUrl)
+
+	// Create the block client with its dependency on the events client.
+	eventsQueryClient := eventsquery.NewEventsQueryClient(flagCometWebsocketUrl)
+	deps := depinject.Supply(eventsQueryClient)
+	blockClient, err := blockclient.NewBlockClient(ctx, deps, flagCometWebsocketUrl)
+	if err != nil {
+		return fmt.Errorf("failed to create block client: %w", err)
+	}
+
+	log.Println("INFO: Creating AppGate server...")
+
+	keyRecord, err := clientCtx.Keyring.Key(flagSigningKey)
+	if err != nil {
+		return fmt.Errorf("failed to get key from keyring: %w", err)
+	}
+
+	appAddress, err := keyRecord.GetAddress()
+	if err != nil {
+		return fmt.Errorf("failed to get address from key: %w", err)
+	}
+	signingAddress := ""
+	if flagSelfSigning {
+		signingAddress = appAddress.String()
+	}
+
+	// Convert the key record to a private key and return the scalar
+	// point on the secp256k1 curve that it corresponds to.
+	// If the key is not a secp256k1 key, this will return an error.
+	signingKey, err := recordLocalToScalar(keyRecord.GetLocal())
+	if err != nil {
+		return fmt.Errorf("failed to convert private key to scalar: %w", err)
+	}
+	signingInfo := appgateserver.SigningInformation{
```

Comment:
I still feel that the signing key construction should be done in the constructor. We should pass in the `signing-key` and `self-sign` flag values and off-load the cmd from any `appGateServer` knowledge except the dependencies.

WDYT?

Reply:
👍🏼